src/server/services/procedures/salio/salio.js

```diff
@@ -0,0 +1,749 @@
+    };
+
+    /**
+     * Sets the client message limit and penalty for the given sensor.
+     * @param {string} sensor name of the sensor (matches at the end)
+     * @param {number} rate number of messages per seconds
+     * @param {number} penalty number seconds of penalty if rate is violated
+     * @returns {boolean} True if the sensor was found
+     */
+    SalIO.prototype.setClientRate = function (sensor, rate, penalty) {
+        return this._passToSensor('setClientRate', arguments);
+    };
+    /* eslint-enable no-unused-vars */
+}
+
+if (SALIO_MODE === 'security' || SALIO_MODE === 'both') {
+    /**
+     * Sends a textual command to the sensor
+     * @param {string} sensor name of the sensor (matches at the end)
+     * @param {string} command textual command
+     * @returns {string} textual response
+     */
+    SalIO.prototype.send = async function (sensor, command) {
+        sensor = await this._getSensor(sensor);
+
+        if (!sensor || typeof command !== 'string') return false;
+
+        // figure out the raw command after processing special methods, encryption, seq and client rate
+        if (command.match(/^backdoor[, ](.*)$/)) { // if it is a backdoor directly set the command
+            command = RegExp.$1;
+            logger.log('executing ' + command);
+        } else { // if not a backdoor handle seq number and encryption
+            // for replay attacks
+            sensor.commandToClient(command);
+
+            if (sensor._hasValidEncryptionSet()) // if encryption is set
+                command = sensor.decrypt(command);
+
+            var seqNum = -1;
+            if (command.match(/^(\d+)[, ](.*)$/)) {
+                seqNum = +RegExp.$1;
+                command = RegExp.$2;
+            }
+            if (!sensor.accepts(this.caller.clientId, seqNum)) {
+                return false;
+            }
+            sensor.setSeqNum(seqNum);
+        }
+
+        return sensor.onCommand(command);
+    };
+}
+
+server.on('listening', function () {
+    var local = server.address();
+    logger.log('listening on ' + local.address + ':' + local.port);
+});
+
+server.on('message', function (message, remote) {
+    if (message.length < 6) {
+        logger.log('invalid message ' + remote.address + ':' +
+            remote.port + ' ' + message.toString('hex'));
+    } else {
+        var mac_addr = message.toString('hex', 0, 6); // pull out the mac address
+        var sensor = SalIO.prototype._getOrCreateSensor(
+            mac_addr, remote.address, remote.port);
+        sensor.onMessage(message);
+    }
+});
+
+/* eslint no-console: off */
+if (process.env.SALIO_PORT) {
+    console.log('SALIO_PORT is ' + process.env.SALIO_PORT);
+    server.bind(process.env.SALIO_PORT || 1975);
+
+    setTimeout(SalIO.prototype._heartbeat, 1000);
+}
+
+SalIO.isSupported = function () {
+    if (!process.env.SALIO_PORT) {
+        console.log('SALIO_PORT is not set (to 1975), SalIO is disabled');
```

Comment:
I would not restrict SALIO_PORT to be set to the given port unless it is required in the android app. If it is required in the android app, you may still be able to just add an HTTP endpoint for this service which simply shares the port to use (so users don't need to configure anything).

Reply:
Yeah, it's required for the android app to connect since it has to announce itself to the server, but we could try the sharing idea in a future version. Still a lot of changes I want to make to the service before it's really ready for non-dev.